File: typescript/base/src/main/java/lang/ProcessBuilder.ts
const childProcess = require("node:child_process");

export class ProcessBuilder {
  private readonly command: string;

  public constructor(commandLine: string) {
    this.command = commandLine;
  }

  public startRead(): any {
    return ProcessBuilder.start(this.command, "r");
  }

  public startWrite(): any {
    return ProcessBuilder.start(this.command, "w");
  }

  public static start(commandLine: string, mode: string): any {
    if (!commandLine || commandLine.length === 0 || !mode || mode.length === 0) {
      return null;
    }

    const stdio = mode === "w"
      ? ["pipe", "inherit", "inherit"]
      : ["inherit", "pipe", "inherit"];

    return childProcess.spawn(commandLine, {
      shell: true,
      stdio
    });
  }

  public static close(processHandle: any): number {
    if (!processHandle) {
      return -1;
    }

    try {
      if (processHandle.stdin && typeof processHandle.stdin.end === "function") {
        processHandle.stdin.end();
      }
      if (typeof processHandle.kill === "function") {
        processHandle.kill();
      }
      return 0;
    }
    catch (_error) {
      return -1;
    }
  }
}

File: cpp/base/src/main/java/lang/ProcessBuilder.cpp
#include <cstdio>

#include "java/lang/ProcessBuilder.h"

namespace java {

ProcessBuilder::ProcessBuilder(const char *commandLine):
    command(commandLine)
{
}

void *
ProcessBuilder::startRead() const {
    return start(command, "r");
}

void *
ProcessBuilder::startWrite() const {
    return start(command, "w");
}

void *
ProcessBuilder::start(const char *commandLine, const char *mode) {
    if ( commandLine == nullptr || commandLine[0] == '\0' || mode == nullptr || mode[0] == '\0' ) {
        return nullptr;
    }
    return static_cast<void *>(popen(commandLine, mode));
}

int
ProcessBuilder::close(void *processHandle) {
    FILE *handle = static_cast<FILE *>(processHandle);
    if ( handle == nullptr ) {
        return -1;
    }
    return pclose(handle);
}

}
